Add can_view_editor_onboarding property to editor session events

dcalhoun opened the pull request:
* **Gutenberg:** https://github.com/WordPress/gutenberg/pull/32739
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3625
* **WPiOS:** https://github.com/wordpress-mobile/WordPress-iOS/pull/16574

Add `can_view_editor_onboarding ` property to editor session events to allow analyzing how capabilities like the editor onboarding impact user behavior. 

To test:
1. Open Tracks to observe editor session events.
1. Uninstall any preexisting WordPress app from device. 
1. Install WordPress app for this PR. 
1. Launch the Gutenberg block editor. 
1. ℹ️ **Expected:** The "Tap to add content" tooltip is displayed. 
1. Close the Gutenberg block editor. 
1. Launch the Gutenberg block editor. 
1. ℹ️ **Expected:** The "Tap to add content" tooltip is _not_ displayed. 
1. Close the Gutenberg block editor. 
1. ℹ️ **Expected:** Four editor session events are captured: 
    * `wpandroid_editor_session_start` with `can_view_editor_onboarding: true`.
    * `wpandroid_editor_session_end` with `can_view_editor_onboarding: true`.
    * `wpandroid_editor_session_start` with `can_view_editor_onboarding: true`.
    * `wpandroid_editor_session_end` with `can_view_editor_onboarding: true`. 

## Regression Notes
1. Potential unintended areas of impact
  Modifications to the post editor analytics and publish ability could introduce regressions to publishing or editor-related analytic events. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manually verified publishing posts worked and related analytic events were triggered for the Gutenberg block editor. 
3. What automated tests I added (or what prevented me from doing so)
  Additional automated tests for this single, experimental property addition are likely not that valuable. 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dcalhoun committed:
Add can_view_editor_onboarding property to editor_session_start event

Property added to allow analyzing how capabilities like the editor
onboarding impact user behavior.

dcalhoun committed:
Add can_view_editor_onboarding property to editor_session_end event

Property added to allow analyzing how capabilities like the editor
onboarding impact user behavior.

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java:
```diff
@@ -94,12 +96,15 @@ public static PostEditorAnalyticsSession getNewPostEditorAnalyticsSession(
             properties.put(KEY_UNSUPPORTED_BLOCKS,
                     unsupportedBlocksList != null ? unsupportedBlocksList : new ArrayList<>());
+            if (capabilities != null) {
+                properties.put(KEY_CAN_VIEW_EDITOR_ONBOARDING, capabilities.getCanViewEditorOnboarding());
```
Is there a more appropriate approach to accessing this value rather than importing the `GutenbergPropsBuilder` module and making this value public?

mchowning replied:
Since you're not making anything in the `GutenbergPropsBuilder` mutable, I think doing it this way is fine. 

With that said, this is not a blocker, but I would do this differently and just pass the relevant boolean value to the `PostEditorAnalyticsSession` instead of the entire `GutenbergPropsBuilder` object. It feels to me like this is kind of using the `GutenbergPropsBuilder` for a purpose (passing information to this class) that it was not designed for (i.e., we wouldn't want someone in the future to add something to the `GutenbergPropsBuilder` just in order to use it in this method). This is admittedly a bit of a matter of style, so I do not consider this a blocker and you can leave this as-is if you prefer, but I would do something like this instead: https://github.com/wordpress-mobile/WordPress-Android/commit/ce5a6245ecb947527a88f6ea2075473bccc2a334.

dcalhoun replied:
Thanks, @mchowning! This is exactly the type of feedback for which I was looking. I am less familiar with Java, so I welcome the ideas. 

> With that said, this is not a blocker, but I would do this differently and just pass the relevant boolean value to the `PostEditorAnalyticsSession` instead of the entire `GutenbergPropsBuilder` object. 

This makes total sense, particularly with the rationale you provided. Passing the entirety of capabilities as the parameter is a leftover from my original implementation where I serialized the entire `capabilities` as a stringified JSON object. However, that value is less valuable in Tracks, as we cannot easily segment on a single capability. 

So, I agree with your changes in https://github.com/wordpress-mobile/WordPress-Android/commit/ce5a6245ecb947527a88f6ea2075473bccc2a334. If you don't mind, I'll likely include them. It may even make sense to update the iOS implementation to mimic its structure as well.

dcalhoun replied:
Refactored this to rely upon a recommended boolean value in https://github.com/wordpress-mobile/WordPress-Android/pull/14766/commits/b690fcd5b738667e9140c49901ee80b154cc0111.

mchowning commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java:
```diff
@@ -148,6 +153,9 @@ public void end() {
             Map<String, Object> properties = getCommonProperties();
             properties.put(KEY_OUTCOME, mOutcome.toString().toLowerCase(Locale.ROOT));
+            if (capabilities != null) {
+                properties.put(KEY_CAN_VIEW_EDITOR_ONBOARDING, capabilities.getCanViewEditorOnboarding());
```
Do we need to send this property when the editor session ends? Won't it just always be `false`?

dcalhoun replied:
> Why do we need to send this property when the editor session ends? 

The hope is that having this capability reported would allow us to observe any differences in user behavior between users with and without this capability enabled. Including the property on for `wpandroid_session_start` and `wpandroid_session_end` was recommended by @hypest. From my perspective, including the property on these two events does make sense, with them serving as "bookends" to other events that occur in the editor.  

> Won't it just always be `false`?

Currently, yes, it will always be `false`. However, I am actively refactoring how the capability is computed so that once a user has this capability enabled, it is always enabled. This involves separating the tracking for `hasLaunchedGutenbergEditor` to no longer be [a part of the capability computation](https://github.com/wordpress-mobile/WordPress-Android/blob/2814689c911c394bab847696dba6d3dfafa94d8b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2296-L2298). 

Knowing this context, do you think it makes sense to add this property to `wpandroid_session_end`?

AmandaRiu replied:
Makes sense to me 👍

mchowning reviewed (commented):
Looks good and is working well David! I left two comments, but only the one about whether we need this property in the "end" event is a "blocker" in the sense that I'd like to understand better why we need to do that before approving. Nice work! 👍

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into update/editor-session-analytic-properties

dcalhoun committed:
Separate computation of onboarding visibility and analytic property

Computing these two values separately enables limiting display of the
editor onboarding to a single time while still allowing us to segment
related analytic events by users who have been included in the cohort.

dcalhoun committed:
Remove unused import

dcalhoun commented:
@mchowning this work has been refactored to follow the approach we discussed in chat. Will you please review at your convenience?

mchowning commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergPropsBuilder.kt:
```diff
@@ -20,7 +20,7 @@ data class GutenbergPropsBuilder(
     private val featuredImageId: Int,
     private val editorTheme: Bundle?,
-    private val canViewEditorOnboarding: Boolean
+    val canViewEditorOnboarding: Boolean
```
This can stay `private` now, right?

dcalhoun replied:
Yes, you are correct. Good catch!

dcalhoun replied:
Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/14766/commits/dcc546e678435e7ee80a2d093074ca83f89d9789.

mchowning commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java:
```diff
@@ -148,6 +150,7 @@ public void end() {
             }
             Map<String, Object> properties = getCommonProperties();
             properties.put(KEY_OUTCOME, mOutcome.toString().toLowerCase(Locale.ROOT));
+            properties.put(KEY_CAN_VIEW_EDITOR_ONBOARDING, canViewEditorOnboarding != null && canViewEditorOnboarding);
```
Do you think the `canViewEditorOnboarding != null &&` part of this might throw off our analytics. This would mean that every time the `end` method is called from a stories post, we'll persist a `false` value (because that class passes `null`), whereas that same user would return a `true` value if they opened and closed Gutenberg?

Same comment for the `start()` method.

dcalhoun replied:
It's difficult to say whether it will throw off analytics, as I'm not sure if/why we would analyze non-Gutenberg (e.g. Story) post analytic events. However, I definitely agree setting it to `false` is less accurate. Good catch. 

I might say we wrap setting the property in a conditional again to avoid setting an inaccurate value. Do you think that would suffice? 
```suggestion
            if (canViewEditorOnboarding != null) {
                properties.put(KEY_CAN_VIEW_EDITOR_ONBOARDING, canViewEditorOnboarding);
            }
```

mchowning replied:
> I might say we wrap setting the property in a conditional again to avoid setting an inaccurate value. Do you think that would suffice?

👍  Sounds good

dcalhoun replied:
Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/14766/commits/335b4662bff4423e5fea328c4346a152315008d1.

mchowning commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -2293,9 +2293,7 @@ private GutenbergPropsBuilder getGutenbergPropsBuilder() {
-        boolean canViewEditorOnboarding = (
-                mAccountStore.getAccount().getUserId() % 100 >= (100 - EDITOR_ONBOARDING_PHASE_PERCENTAGE)
-                || BuildConfig.DEBUG) && !AppPrefs.hasLaunchedGutenbergEditor();
+        boolean canViewEditorOnboarding = canViewEditorOnboarding() && !AppPrefs.hasLaunchedGutenbergEditor();
```
What would you think about renaming either this variable (and all the downstream variables) or the `canViewEditorOnboarding()` function so that they're different since they represent different things? Having them use the same name feels like something that will result in a bug sometime in the future to me. At a minimum, it feels confusing to me.

dcalhoun replied:
I agree. I began to rename the capability itself within `GutenbergPropsBuilder` to be `editorOnboarding`, but paused to avoid needing to update Gutenberg itself. That said, if you also feel it is worthwhile, then I can make that change now.

mchowning replied:
I think it is worthwhile. Perhaps it would work to just rename the function to something like `eligibleForEditorOnboarding()` to avoid having to update Gutenberg.

dcalhoun replied:
Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/14766/commits/55d78c8d5bbe1d9be8ce0f10e7f9cf5af22d187c.

mchowning reviewed (commented):
Tested, and it is working as expected (all `true`s are getting logged). 👍

dcalhoun committed:
Privatize editor onboarding value

There is no longer a need to access this value directly.

dcalhoun committed:
Avoid inaccurate set of editor onboarding property

Previously, a false value would be set whenever the editor onboarding
value was null, which could've produced misleading analytic events.
Avoiding setting the property unless it is defined is likely a better
approach.

dcalhoun committed:
Rename capability for clarity

The capability property passed to Gutenberg is dependent upon both the
percentage eligibility and whether the editor has been launched
previously. Renaming the property may help convey the multiple factors.

dcalhoun commented:
@mchowning I believe I have addressed all the great comments you provided. Will you please review this PR and the sibling https://github.com/wordpress-mobile/gutenberg-mobile/pull/3625 and https://github.com/WordPress/gutenberg/pull/32739 at your earliest convenience? Thank you! 🙇🏻

AmandaRiu reviewed (approved):
@dcalhoun reviewed the code, tested the changes, and verified the tracks events. Looks good! :shipit:

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into update/editor-session-analytic-properties

dcalhoun committed:
Update gutenbergMobileVersion ref
